features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericGraph.java

```diff
@@ -192,12 +197,20 @@ public void addEdge(GenericEdge edge) {
         Objects.requireNonNull(edge);
         Objects.requireNonNull(edge.getId());
         if (jungGraph.containsEdge(edge)) return; // already added
-        if (edge.getSource().getNamespace().equalsIgnoreCase(getNamespace()) && getVertex(edge.getSource().getId()) == null) {
-            addVertex((GenericVertex) edge.getSource());
-        }
-        if (edge.getTarget().getNamespace().equalsIgnoreCase(getNamespace()) && getVertex(edge.getTarget().getId()) == null) {
-            addVertex((GenericVertex) edge.getTarget());
+        if(!this.getNamespace().equals(edge.getNamespace())){
+            throw new IllegalArgumentException(
+                    String.format("The namespace of the edge (%s) doesn't match the namespace of this graph (%s). Edge: %s ",
+                    edge.getNamespace(), this.getNamespace(), edge.toString()));
         }
+        // TODO patrick: discuss with mvr: the following casting won't work anymore since we are operating on actual Vertexrefs.
```

Comment:
As an edge consists of two vertices, the idea was to add the vertices if they are not part of the graph yet, as they need to be added anyways. However, as an edge now only consists of `VertexRef`s, either source or target must be of type `Vertex` to add it to the Graph. To resolve a `VertexRef` it must already be known by the Graph. For now I'd say we have to ensure that the graph already knows all vertex refs of the same namespace, instead of source AND target.

Reply:
done